### ISSUE TITLE
riscvOVPsimPlus P Extension

### ISSUE THREAD

**Issue by fpedd** [riscv-ovpsim/imperas-riscv-tests#21]
According to section 1.7 of the manual (`riscvOVPsimPlus` `User Guide.pdf` `Version 0.8` `Created 27 May 2022`) the riscvOVPsimPlus simulator should support the `RISC-V P DSP/SIMD Extension (dsp version)`:
- `Version 0.5.2 : DSP Architecture Version 0.5.2`
- `Version 0.9.6 : DSP Architecture Version 0.9.6`

However, I am getting:
```
Warning (RISCV_UDEC) CPU 'riscvOVPsim/cpu' 0x00010646 40c787f7 undef: Undecoded instruction
Processor Exception (PC_PRX) Processor 'riscvOVPsim/cpu' 0x0: 0000     illegal
Processor Exception (PC_FIX) Executing at uninitialized address 0x0
```
from the first packed P instruction, the simulator encounters:
```
   10646:	40c787f7          	add16	a5,a5,a2
```

Running the same code on Spike with P support succeeds just fine. 

I supplied `--override riscvOVPsim/cpu/add_Extensions=MAFDCBP`, as well as `--variant RVB32I`, and I can observe, that the 15th bit (corresponding to P) inside the `misa` reg is set: 00000000 00000000 **1**0010001 00101101. Removing `P` resets this bit.  

I am running
```
riscvOVPsimPlus (64-Bit) v20220527.0 Open Virtual Platform simulator from www.IMPERAS.com.
```

PS: I am also required to turn on the `B` extension, since I need instructions from [Zbpbo](https://github.com/riscv/riscv-p-spec/blob/master/P-ext-proposal.adoc#zbpbo). OVPsim is also confirming this on the terminal:
```
Info   Type                  : riscv (RVB32I+ABCDFMP)
```
While OVPsim happily eats `bitmanip` instructions now, the 1st (B) bit in `misa` is not set: 00000000 00000000 10010001 001011**0**1. Why would it not be set?

**Comment by duncangraham-Imperas** [comment]
Looks like you have correctly enabled the extension, however, the default bitmanip version supported by riscvOVPsimPlus RISCV processor model is v1.0.0.
You also neeed to set the configuration parameter bitmanip_version? Unlike Spike, that I understand is hardcoded for the version of an extension at the time of creation, the OVP RISCV model (included with riscvOVPsimPlus, supports all extension versions. The default is the current at time of release but all others can be selected.

If you use the command line argument -showoverrides you can see all the available parameter overrides, for example
--override riscvOVPsim/cpu/bitmanip_version=1.0.0

NOTE: If you set the parameter to an illegal value , it will provide you all the possible legal values
--override riscvOVPsim/cpu/bitmanip_version=illegal
Error (OP_IU) Setting enumerated parameter 'riscvOVPsim/cpu/bitmanip_version' to the illegal value 'illegal'. Legal values are:
    0.90
    0.91
    0.92
    0.93-draft
    0.93
    0.94
    1.0.0
    master

**Comment by fpedd** [comment]
Okay, thanks. But this does not solve my P extension problem, right?

**Comment by duncangraham-Imperas** [comment]
Sorry, missed you were talking about B and P-DSP in the issue .. I saw talk of versions and bitmanip at the end of the question .. and off I went. Let me try again ;-)

Bitmanip: When the B extension was split into several sub-extensions (Zba, Zbb etc.) the specification also changed such that the presence of the B extension is no longer indicated in the MISA. I think this was around version 0.92 to 0.93 transition.

DSP: The DSP versions supported are, as you say, 0.5.2 (default) and 0.9.6. What version do you expect that the binary encoding 0x40c787f7 decodes? and what is the encoding you are expecting?
Did you use the parameter override of dsp_version to try version 0.9.6 with this encoding.

**Comment by jstraus59** [comment]
The important thing to be aware of here is that instruction encodings can change between pre-ratification versions of RiscV extensions.

For DSP Versions 0.5.2 and earlier, bits 6:0 of opcode are b'1111111 while in later versions they are b'1110111. The encoding of your instruction indicates that it is using the post-0.5.2 encodings.

riscvOVPsim by default uses the 0.5.2 encodings, as can be seen by running:
```
$ ./riscvOVPsimPlus.exe --override riscvOVPsim/cpu/add_Extensions=P --showoverrides | grep dsp_version
    --override riscvOVPsim/cpu/dsp_version=0.5.2 (Enumeration) (default=0.5.2) (default) Specify required DSP Architecture version
```
So you must use an override to select the 0.9.6 encodings:
```
$ ./riscvOVPsimPlus.exe --override riscvOVPsim/cpu/add_Extensions=P --override riscvOVPsim/cpu/dsp_version=0.9.6 ...
```

**Comment by fpedd** [comment]
@duncangraham-Imperas and @jstraus59 thank you for your help. `--override riscvOVPsim/cpu/dsp_version=0.9.6` is exactly what I was missing. It works now! :)

Would not have thought that the simulator was defaulting to the earlier version. Also, `--override riscvOVPsim/cpu/dsp_version` does not seem to be documented in the manuals I received in the riscvOVPsimPlus bundle.
